### PR TITLE
Classify movement

### PR DESCRIPTION
**Pull request by gferraro**
New model uses a different input format, so added support for that.
Moved classification of tracks into keras class for new models
Moved preprocess from dataset into it's own class and added new functions
Ensured values are floats instead of float32 so metadata saving works

Have tested against a cptv file using the original model, and the new keras frame model and get the same accuracy and clarity as current master branch as expected.

### CODE REVIEW

File: ml_tools/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+    channel=TrackChannels.filtered,
+    require_movement=False,
+):
+    """Return 2 images describing the movement, one has dots representing
+    the centre of mass, the other is a collage of all frames
+    """
+
+    i = 0
+    if dim is None:
+        # gp should be from track data
```
Comment (clare) [inline ml_tools/preprocess.py:258]: Should this still be here?

File: ml_tools/preprocess.py [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+    if max == min:
+        if max == 0:
+            return np.zeros((data.shape)), False
+        return data / max, False
+    data -= min
+    data = data / (max - min) * new_max
+    return data, True
+
+
+def movement_images(
```
Comment (clare) [inline ml_tools/preprocess.py:245]: This is quite confusing having both the overlay and the dots code together.   Maybe separate it into two functions?  Then we can explore models with one or other or both the extra movement images.   We don't want to be caulculating something that doesn't give us much extra value.

File: ml_tools/preprocess.py [hunk 3]
```diff
@@ -0,0 +1,332 @@
...
+    min_distance = 2
+    for i, frame in enumerate(frames):
+        region = regions[i]
+
+        x = int(region.mid_x)
+        y = int(region.mid_y)
+
+        if prev is not None:
+            distance = math.sqrt(pow(prev[0] - x, 2) + pow(prev[1] - y, 2))
+            center_distance += distance
```
Comment (clare) [inline ml_tools/preprocess.py:279]: The centre distance calculation doesn't calculate disatnce from last image overlay.   Eg consider when the animal moves slightly one way and then back again.   You really want to calculate the the distance from the previous overlay not the previous frame.   

Also math.sqrt is expensive.   This is needed for drawing the lines, but not for calculating whether the animal has moved far enough away to draw another overlay.   So it should really be any more expensive.

**Review by clare — APPROVED** [review]
Wow that was a lot to review.   I have a few concerns about the way the overlay image is calculated.   Also I think we need to start documenting how the model works.   Can you put up an example of what the movement images look like in the read me maybe?

File: ml_tools/kerasmodel.py [hunk 4]
```diff
@@ -1,10 +1,15 @@
...
-from ml_tools.dataset import TrackChannels
+from track.track import TrackChannels
+from classify.trackprediction import TrackPrediction
+from ml_tools.preprocess import (
+    preprocess_movement,
+    preprocess_frame,
+)
 
 
 class KerasModel:
```
[the remark this answers is not in the document]
Comment (gferraro) [reply ml_tools/kerasmodel.py:15]: its just that it's tensorflow v2 keras framework rather than based of 1.15 as the old one, I imagine soon we can get rid of the old model code if we are happy with this

File: ml_tools/kerasmodel.py [hunk 5]
```diff
@@ -188,40 +215,93 @@ def classify_frame(self, frame, preprocess=True):
...
-    data = data / (max - min)
-    np.clip(data, a_min=0, a_max=None, out=data)
+        frames_per_classify = self.square_width ** 2
+        num_frames = len(data)
 
-    data = data[np.newaxis, :]
-    data = np.transpose(data, (1, 2, 0))
-    data = np.repeat(data, output_dim[2], axis=2)
-    data = reisze_cv(data, output_dim)
+        # note we can use more classifications but since we are using all regions
```
[the remark this answers is not in the document]
Comment (gferraro) [reply ml_tools/kerasmodel.py:278]: yep thats exactly what i'm doing, will update comment

File: ml_tools/preprocess.py [hunk 6]
```diff
@@ -0,0 +1,332 @@
...
+        dim=square.shape,
+        channel=channel,
+        require_movement=True,
+    )
+    dots = dots / 255
+    overlay, success = normalize(overlay, min=0)
+    if not success:
+        return None
+
+    data = np.empty((square.shape[0], square.shape[1], 3))
```
[the remark this answers is not in the document]
Comment (gferraro) [reply ml_tools/preprocess.py:219]: yes but if we decide to use a rectangle. this will continue to work. and square_width is actually how many frames to tile accross, will think of a better name

File: ml_tools/preprocess.py (hunk 1 above)
Comment (gferraro) [reply ml_tools/preprocess.py:258]: yes and no, i'm going to make dim mandatory parameter and remove this

File: ml_tools/preprocess.py (hunk 3 above)
Comment (gferraro) [reply ml_tools/preprocess.py:279]: Good point, i will have to change the overlay calculation and check it doesn't ruin the model accuracy.

Yep will remove math.sqrt

File: ml_tools/preprocess.py [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+from track.track import TrackChannels
+
+
+# size to scale each frame to when loaded.
+FRAME_SIZE = 48
+
+MIN_SIZE = 4
+
+
+def preprocess_segment(
```
[the remark this answers is not in the document]
Comment (gferraro) [reply ml_tools/preprocess.py:16]: good point added a new file

File: ml_tools/kerasmodel.py [hunk 8]
```diff
@@ -188,40 +215,93 @@ def classify_frame(self, frame, preprocess=True):
...
+        num_classifies = math.ceil(float(num_frames) / frames_per_classify)
+        frame_sample = np.arange(num_frames)
+        np.random.shuffle(frame_sample)
+        for i in range(num_classifies):
+            seg_frames = frame_sample[:frames_per_classify]
+            if len(seg_frames) == 0:
+                break
+            segment = []
+            # update remaining
+            frame_sample = frame_sample[frames_per_classify:]
```
[the remark this answers is not in the document]
Comment (gferraro) [reply ml_tools/kerasmodel.py:276]: wishfull thinking :)
